src/sentry/spans/consumers/process/factory.py

```diff
@@ -301,38 +186,196 @@ def create_with_partitions(
         commit: Commit,
         partitions: Mapping[Partition, int],
     ) -> ProcessingStrategy[KafkaPayload]:
+        committer = CommitOffsets(commit)
 
-        produce_step = Produce(
-            producer=self.producer,
-            topic=self.output_topic,
-            next_step=NoOp(),
-        )
+        buffer = RedisSpansBufferV2(assigned_shards=[p.index for p in partitions])
 
-        unfold_step = Unfold(generator=expand_segments, next_step=produce_step)
-
-        commit_step = CommitSpanOffsets(commit=commit, next_step=unfold_step)
-
-        batch_processor = RunTask(
-            function=batch_write_to_redis,
-            next_step=commit_step,
+        flusher = SpanFlusher(
+            buffer,
+            self.producer,
+            self.output_topic,
+            self.max_flush_segments,
+            self.max_inflight_segments,
+            next_step=committer,
         )
 
-        batch_step = BatchStep(
+        run_task = run_task_with_multiprocessing(
+            function=partial(process_batch, buffer),
+            next_step=flusher,
             max_batch_size=self.max_batch_size,
             max_batch_time=self.max_batch_time,
-            next_step=batch_processor,
+            pool=self.__pool,
+            input_block_size=self.input_block_size,
+            output_block_size=self.output_block_size,
         )
 
-        return run_task_with_multiprocessing(
-            function=process_message,
-            next_step=batch_step,
+        batch = BatchStep(
             max_batch_size=self.max_batch_size,
             max_batch_time=self.max_batch_time,
-            pool=self.__pool,
-            input_block_size=self.input_block_size,
-            output_block_size=self.output_block_size,
+            next_step=run_task,
         )
 
+        def add_timestamp_cb(message: Message[KafkaPayload]) -> tuple[int, KafkaPayload]:
```

Comment:
_nit_: Clarify which timestamp. Please also consider describing the rationale why we use the _produce_ timestamp rather than receive or span timestamps in a comment.

```suggestion
        def add_produce_timestamp_cb(message: Message[KafkaPayload]) -> tuple[int, KafkaPayload]:
```